api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -182,99 +187,111 @@ message HttpTrailers {
-  // If set, manipulate HTTP headers along with the body. Note that
-  // this is only possible if the headers have not been sent down
-  // the filter chain yet. Practically, then, this field will be ignored
-  // unless the request or response BodySendMode is BUFFERED or
-  // BUFFERED_PARTIAL, which means that Envoy is buffering the body
-  // in memory while waiting for this response.
-  HeaderMutation header_mutation = 1;
+  CommonResponse response = 1;
+}
 
-  // Request that the last body chunk delivered be modified.
-  BodyMutation body_mutation = 2;
-
-  // If set, add trailers to the message. If and only if this
-  // field is combined with the end_stream flag, then
-  // the trailers will be sent along with the request or
-  // response -- otherwise, this field is ignored.
-  config.core.v3.HeaderMap trailers = 3;
-
-  // If true, then replace the request and response with the contents
-  // of this message. If header_mutation is set and headers have not
-  // already been forwarded down the filter chain, apply it to the
-  // headers. If body_mutation is set and contains a body, then add that
-  // body to the request or response, even if one does not already exist --
-  // otherwise, clear the body. Any additional body and trailers
-  // received from downstream will be ignored, and any additional messages
-  // from the remote server on either the request or response path will
-  // also be ignored for the duration of this request or response.
-  bool end_of_stream = 4;
+// This message contains common fields between header and body responses.
+// [#next-free-field: 6]
+message CommonResponse {
+  enum ResponseStatus {
+    // Apply the mutation instructions in this message to the
+    // request or response, and then continue processing the filter
+    // stream as normal. This is the default.
+    CONTINUE = 0;
+
+    // Replace the request or response with the contents
+    // of this message. If header_mutation is set, apply it to the
+    // headers. If body_mutation is set and contains a body, then add that
+    // body to the request or response, even if one does not already exist --
+    // otherwise, clear the body. Any additional body and trailers
+    // received from downstream will be ignored, and any additional messages
+    // from the remote server on either the request or response path will
+    // also be ignored for the duration of this request or response.
+    // This can be used to add a body to a request or response that does not
+    // have one already.
+    CONTINUE_AND_DO_NOT_END_STREAM = 1;
+
+    // Replace the request or response as with CONTINUE_AND_DO_NOT_END_STREAM,
+    // but do not continue filter processing yet -- the remote server can
+    // send additional ProcessingResponse messages with this value
+    // set that further modify the message or add new data, and
+    // must eventually reply with one of the "CONTINUE" statuses
+    // in order for filter processing to resume.
```

Comment:
I would make it clear that this results in buffering? Stepping back, is there any real reason that we would ever stream chunks from the ext proc server and buffer it in the filter? Can this all be simplified into either CONTINUE or CONTINUE_AND_DO_NOT_END_STREAM?

Reply:
I think in the long conversation that we had last week, we talked about a case -- adding a body to a message with no body already -- in which the filter chain allows a filter to insert data to the filter chain, and there's that specific return value from the "on*Headers" callback that's used for it. So this return value supports that.

However, given that the filter can't actually return that special error code, I'd certainly rather simplify this thing by only having two states -- continue and continue_and_do_not_end_stream, because the ext_proc server can always buffer. Then of course we can add another one in the future.

If you're OK with that I'll make it.

Reply:
Yes, I think this makes the most sense for now.